tests/cecil-tests/ApiCapitalizationTest.cs

```diff
@@ -0,0 +1,223 @@
+		};
+
+
+		Dictionary<string, HashSet<string>> allowedMethods = new Dictionary<string, HashSet<string>> () {
+			["Dlfcn"] = new HashSet<string> () { "dlopen", "dlerror", "dlsym" }
+		};
+
+		public bool IsSkip (string type, string memberName, Dictionary<string, HashSet<string>> allowed)
+		{
+			if (allowed.TryGetValue (type, out var result)) {
+				if (result.Contains (memberName))
+					return true;
+			}
+
+			return Char.IsUpper (memberName [0]);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void PropertiesCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				var typeName = type.Name;
+				return type.Properties
+						.Where (p => p.GetMethod?.IsPublic == true || p.SetMethod?.IsPublic == true)
+						.Where (p => !IsSkip (type.Name, p.Name, allowedProperties) && !IsMemberObsolete (p) && !p.IsSpecialName)
+						.Select (p => p.Name);
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void MethodsCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				return from m in type.Methods
+						where m.IsPublic && !IsSkip (type.Name, m.Name, allowedMethods) && !IsUnique (assemblyPath, m)
+						select m.Name;
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void EventsCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				return from e in type.Events
+					   where !(Char.IsUpper (e.Name [0]))
+					   select e.Name;
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void FieldsCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				return from f in type.Fields
+					   where f.IsPublic && f.IsFamilyOrAssembly && !IsSkip (type.Name, f.Name, allowedFields)
+					   select f.Name;
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		public void CapitalizationTest (string assemblyPath, Func<TypeDefinition, IEnumerable<string>> selectLambda)
+		{
+
+			if (assemblyCache.TryGetValue (assemblyPath, out var cache)) {
+				var typeDict = new Dictionary<string, string> ();
+
+				var publicTypes = cache.MainModule.Types.Where ((t) => t.IsPublic && !IsMemberObsolete (t));
+
+				foreach (var type in publicTypes) {
+					TypeCheck(type, selectLambda, typeDict);
+					if (type.HasNestedTypes) {
+						foreach (var nestedType in type.NestedTypes) {
+							TypeCheck (nestedType, selectLambda, typeDict);
```

Comment:
Nested types can also have nested types... it'll work recursively if you move the HasNestedTypes check and corresponding loop into the TypeCheck method.